pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go

```diff
@@ -705,14 +705,21 @@ func constructServicesFromWorkloadEntry(p *networkingv1alpha3.WorkloadEntry, ser
 			// Named targetPort has different semantics from Service vs ServiceEntry
 			if svc.Source.Kind == kind.Service {
 				// Service has explicit named targetPorts.
-				if named, f := svc.PortNames[int32(port.ServicePort)]; f && named.TargetPortName != "" {
-					// This port is a named target port, look it up
-					tv, ok := p.Ports[named.TargetPortName]
-					if !ok {
-						// We needed an explicit port, but didn't find one - skip this port
-						continue
+				if named, f := svc.PortNames[int32(port.ServicePort)]; f {
+					// Target port name should be respected first. Otherwise, fallback to port name
+					if named.TargetPortName != "" {
+						tv, ok := p.Ports[named.TargetPortName]
+						if !ok {
+							// We needed an explicit port, but didn't find one - skip this port
+							continue
+						}
+						targetPort = tv
+					} else {
+						tv, ok := p.Ports[named.PortName]
```

Comment:
why should match portName? I donot see k8s match it, only TargetPortName is matched with pod

Reply:
from the WorkloadEntry [docs](https://istio.io/latest/docs/reference/config/networking/workload-entry/#:~:text=it%20must%20be%20a%20map%20of%20servicePortName%20to%20this%20endpoint%E2%80%99s%20port%2C%20such%20that%20traffic%20to%20the%20service%20port%20will%20be%20forwarded%20to%20the%20endpoint%20port%20that%20maps%20to%20the%20service%E2%80%99s%20portName.): 

>  it must be a map of servicePortName to this endpoint’s port, such that traffic to the service port will be forwarded to the endpoint port that maps to the service’s portName.

Reply:
Have seen it. Sidecar match ServicePort name first then targetPort name. Here it is opposite, we should keep consistent